packages/vechain-kit/src/components/LegalDocumentsModal/LegalDocumentsContent.tsx

```diff
@@ -0,0 +1,196 @@
+import { StickyHeaderContainer } from '@/components/common';
+import { useLegalDocuments, useVeChainKitConfig } from '@/providers';
+import { EnrichedLegalDocument, LegalDocumentSource } from '@/types';
+import {
+    Button,
+    ModalBody,
+    ModalFooter,
+    ModalHeader,
+    Stack,
+    Text,
+    VStack,
+} from '@chakra-ui/react';
+import { useMemo } from 'react';
+import { useForm } from 'react-hook-form';
+import { useTranslation } from 'react-i18next';
+import { Fragment } from 'react';
+
+import { LegalDocumentItem } from './Components';
+
+type Props = {
+    onAgree: (
+        documents: EnrichedLegalDocument | EnrichedLegalDocument[],
+    ) => void;
+    onReject: () => void;
+};
+
+export const LegalDocumentsContent = ({ onAgree, onReject }: Props) => {
+    const { t } = useTranslation();
+    const { darkMode: isDark } = useVeChainKitConfig();
+    const {
+        legalDocuments: { documentsNotAgreed, getDocumentId },
+    } = useLegalDocuments();
+
+    // Separate required and optional documents
+    const requiredDocuments = useMemo(() => {
+        return documentsNotAgreed.filter((document) => document.required);
+    }, [documentsNotAgreed]);
+
+    const optionalDocuments = useMemo(() => {
+        return documentsNotAgreed.filter((document) => !document.required);
+    }, [documentsNotAgreed]);
+
+    // Filter for VechainKit required documents
+    const vechainKitRequiredDocuments = useMemo(() => {
+        return requiredDocuments.filter(
+            (document) =>
+                document.documentSource === LegalDocumentSource.VECHAIN_KIT,
+        );
+    }, [requiredDocuments]);
+
+    // Filter for other required documents
+    const appRequiredDocuments = useMemo(() => {
+        return requiredDocuments.filter(
+            (document) => document.documentSource === LegalDocumentSource.OTHER,
+        );
+    }, [requiredDocuments]);
+
+    // Filter for optional documents
+    const allOptionalDocuments = useMemo(() => {
+        return optionalDocuments;
+    }, [optionalDocuments]);
+
+    const defaultFormValues = documentsNotAgreed.reduce((acc, document) => {
+        acc[getDocumentId(document)] = document.required;
+        return acc;
+    }, {} as Record<string, boolean>);
+
+    const {
+        handleSubmit,
+        register,
+        formState: { isValid },
+    } = useForm<Record<string, boolean>>({
+        defaultValues: defaultFormValues,
+    });
+
+    const onSubmit = (data: Record<string, boolean>) => {
+        const agreedInputs = Object.entries(data)
+            .filter(([_, checked]) => checked)
+            .filter(Boolean);
+        const agreedDocuments = documentsNotAgreed.filter((document) =>
+            agreedInputs.some(([documentId]) => documentId === document.id),
+        );
+
+        if (agreedDocuments.length > 0) {
+            return onAgree(agreedDocuments);
+        }
+    };
```

Comment:
on a similar note, I see more optimisation opportunities here. We could consolidate the multiple filter ops into a single reduce for doc organisation (O(n) instead of O(n×5)), and use a Set in the submit function for O(1) lookups instead of nested array filtering. Happy to show you how if you're interested!